src/chrony.c

```diff
@@ -0,0 +1,1147 @@
+ * 
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ * 
+ **********************************************************************
+*/
+/* Formated using 'GNU indent -bli0 -ce -ppi 2' and some manual alignment
+ * TODO:
+ *	- More robust udp parsing (using offsets instead of structs?)
+ *	  -> Currently chrony parses its data the same way as we do (using structs)
+ *	- Plausibility checks on values received
+ *	  -> Done at higher levels
+ */
+
+#include "config.h"
+
+#if HAVE_SYS_TYPES_H
+#  include <sys/types.h>        /* getaddrinfo */
+#endif
+#if HAVE_SYS_SOCKET_H
+#  include <sys/socket.h>
+#endif
+#if HAVE_NETDB_H
+#  include <netdb.h>
+#endif
+#if HAVE_ARPA_INET_H
+#  include <arpa/inet.h>        /* ntohs/ntohl */
+#endif
+
+#include "collectd.h"
+#include "common.h"             /* auxiliary functions */
+#include "plugin.h"             /* plugin_register_*, plugin_dispatch_values */
+
+#define CONFIG_KEY_HOST    "Host"
+#define CONFIG_KEY_PORT    "Port"
+#define CONFIG_KEY_TIMEOUT "Timeout"
+
+#define URAND_DEVICE_PATH  "/dev/urandom"       /* Used to initialize seq nr generator */
+#define RAND_DEVICE_PATH   "/dev/random"        /* Used to initialize seq nr generator (fall back) */
+
+static const char *g_config_keys[] = {
+  CONFIG_KEY_HOST,
+  CONFIG_KEY_PORT,
+  CONFIG_KEY_TIMEOUT
+};
+
+static int g_config_keys_num = STATIC_ARRAY_SIZE(g_config_keys);
+static int g_is_connected = 0;
+static int g_chrony_socket = -1;
+static time_t g_chrony_timeout = -1;
+static char *g_chrony_host = NULL;
+static char *g_chrony_port = NULL;
+static char *g_plugin_instance = NULL;
+static uint32_t g_chrony_rand = 1;
+static uint32_t g_chrony_seq_is_initialized = 0;
+
+#define PLUGIN_NAME_SHORT "chrony"
+#define PLUGIN_NAME       PLUGIN_NAME_SHORT " plugin"
+#define DAEMON_NAME       PLUGIN_NAME_SHORT
+#define CHRONY_DEFAULT_HOST "localhost"
+#define CHRONY_DEFAULT_PORT "323"
+#define CHRONY_DEFAULT_TIMEOUT 2
+
+/* Return codes (collectd expects non-zero on errors) */
+#define CHRONY_RC_OK    0
+#define CHRONY_RC_FAIL  1
+
+/* Chronyd command packet variables adapted from chrony/candm.h (GPL2) */
+#define PROTO_VERSION_NUMBER 6
+#define IPADDR_UNSPEC 0
+#define IPADDR_INET4  1
+#define IPADDR_INET6  2
+#define IPV6_STR_MAX_SIZE (8*4+7+1)
+
+typedef enum
+{
+  PKT_TYPE_CMD_REQUEST = 1,
+  PKT_TYPE_CMD_REPLY   = 2
+} ePacketType;
+
+typedef enum
+{
+  REQ_N_SOURCES    = 14,
+  REQ_SOURCE_DATA  = 15,
+  REQ_TRACKING     = 33,
+  REQ_SOURCE_STATS = 34
+} eDaemonRequests;
+
+
+typedef enum
+{
+  RPY_NULL             = 1,
+  RPY_N_SOURCES        = 2,
+  RPY_SOURCE_DATA      = 3,
+  RPY_MANUAL_TIMESTAMP = 4,
+  RPY_TRACKING         = 5,
+  RPY_SOURCE_STATS     = 6,
+  RPY_RTC              = 7
+} eDaemonReplies;
+
+#define ATTRIB_PACKED __attribute__((packed))
+typedef struct ATTRIB_PACKED
```

Comment:
Fair enough. Ideally we detect if __attribute__((packed)) is supported by the compiler (Sun Studio has it too iirc). We should do that for the other attributes we use in the codebase to, but let's leave that for another day :)

It's probably good to break compilation, you might even turn the warning into an error, so we get reports for other compilers.

One last thing, clang also defines __GNUC__ so there's no need to test for that seperately.